File: Arthur/Assets/Tests/ApiClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using BestHTTP;
using BestHTTP.Decompression.Zlib;
using BestHTTP.Forms;
using BestHTTP.JSON;
using UnityEngine;
using Random = System.Random;

namespace Tests
{
    public class LastRequest
    {
        public string EndPoint;
        public RequestData RequestData;
    }
    
    public class ApiClient
    {
        // service
        private EncryptionHelper _encryptionHelper = new EncryptionHelper();
        MD5 md5 = MD5.Create();
        public LastRequest LastRequest;
        public RespondData RespondData;
        
        // variables
        public long nonce = 1;
        public string Host;
        public string DeviceId;
        public string Version = "1.0";
        public string VersionKey = "any-version-name";
        public int Timeout;
        public bool isSending = false;
        
        // for testing
        public bool remoteSendInvalidPayload = false;
        public bool localSendInvalidSignBase64 = false;
        public bool localSendInvalidRequestData = false;
        public bool remoteTimeout = false;
        
        // return from server
        public User user;
        public Card[] cards;
        public Card card;
        public long receivedTimestamp;
        public long serverTimestamp;
        public string token = "";
        
        public ApiClient(string host, string deviceId, int timeout)
        {
            Host = host;
            DeviceId = deviceId;
            Timeout = timeout;
            var random = new Random();
            nonce = random.Next(0, 1 << 30);
        }

        public long CurrentTimestamp
        {
            get { return DateTimeOffset.UtcNow.ToUnixTimeSeconds(); }
        }
        
        public long PredictedServerTimestamp
        {
            get { return serverTimestamp + (CurrentTimestamp - receivedTimestamp);  }
        }
        
        
        public bool CanResend
        {
            get { return LastRequest != null && !isSending; }
        }

        public void Resend()
        {
            if (CanResend)
            {
                _request(LastRequest.EndPoint, LastRequest.RequestData);
            }
        }

        public void Login()
        {
            var requestData = new RequestData() {deviceId = DeviceId};
            _request("/users/login", requestData);
        }

        public void ListCards()
        {
            var requestData = new RequestData() {};
            _request("/cards/list", requestData);
        }

        public void CreateCard(int monsterid)
        {
            var requestData = new RequestData() {monsterId = monsterid, cacheKey = UnityEngine.Random.value.ToString()};
            _request("/cards/create", requestData);
        }
        
        public void UpdateCard(int cardId, int exp)
        {
            var requestData = new RequestData() {cardId = cardId, exp = exp, cacheKey = UnityEngine.Random.value.ToString()};
            _request("/cards/update", requestData);
        }
        
        public void DeleteCard(int cardId)
        {
            var requestData = new RequestData() {cardId = cardId, cacheKey = UnityEngine.Random.value.ToString()};
            _request("/cards/delete", requestData);
        }

        public string getQuery(Dictionary<string, string> dict)
        {
            string query = string.Join("&", dict.Select(x => x.Key + "=" + Uri.EscapeDataString(x.Value)).ToArray());
            return query;
        }

        public void _request(string endPoint, RequestData requestData)
        {
            isSending = true;

            requestData.version = Version;
            requestData.versionKey = VersionKey;
            requestData.deviceId = DeviceId;
            requestData.timestamp = PredictedServerTimestamp;
            requestData.session = user.session;
            
            
            Byte[] payload = _encryptionHelper.Encrypt(Encoding.UTF8.GetBytes(localSendInvalidRequestData ? "_" : "" + JsonUtility.ToJson(requestData)), ++nonce);
            string payloadBase64 = Convert.ToBase64String(payload);
            var postForm = new RawJsonForm();
            postForm.AddField("payloadBase64", payloadBase64);
            
            // signature
            Byte[] digest = md5.ComputeHash(payload);
            Byte[] nonceByte = BitConverter.GetBytes(nonce);
            Byte[] signMessage = nonceByte.Concat(digest).ToArray();
            string signedBase64 = Convert.ToBase64String(_encryptionHelper.SignMessage(signMessage));
            
            // query
            Dictionary<string, string> queryDict = new Dictionary<string, string>();
            queryDict.Add("signedBase64", localSendInvalidSignBase64 ? "_" : "" + signedBase64);
            if(token != "")
                queryDict.Add("token", token);
            if(remoteTimeout)
                queryDict.Add("remoteTimeout", "true");
            if(remoteSendInvalidPayload)
                queryDict.Add("remoteSendInvalidPayload", "true");
            var query = getQuery(queryDict);
            
            
            LastRequest = new LastRequest()
            {
                EndPoint = endPoint,
                RequestData = requestData
            };
            
            // send request
            var url = $"{Host}{endPoint}?{query}";
            HTTPRequest request = new HTTPRequest(new Uri(url), HTTPMethods.Post, (originalRequest, response) =>
            {
                if (response != null && response.IsSuccess)
                {
                    if (response.Headers.ContainsKey("content-encoding") && response.Headers["content-encoding"].Contains("encrypted"))
                    {
                        byte[] decrypted = _encryptionHelper.Decrypt(response.Data, nonce);
                        if(decrypted.Length > 10)
                        {
                            RespondData = JsonUtility.FromJson<RespondData>(
                                Encoding.UTF8.GetString(Decompress(decrypted)));


                            receivedTimestamp = RespondData.timestamp;
                            serverTimestamp = RespondData.timestamp;

                            if (RespondData.timestamp - requestData.timestamp > Timeout)
                                RespondData.errorCode = ErrorCode.Timeout;
                            else
                            {
                                user = RespondData.user;
                                if (Mathf.Abs(PredictedServerTimestamp - receivedTimestamp) > 3600)
                                    RespondData.errorCode = ErrorCode.InvalidTimestamp;
                                else
                                {

                                    cards = RespondData.body.cards;
                                    card = RespondData.body.card;
                                }

                                if (!String.IsNullOrEmpty(RespondData.token))
                                    token = RespondData.token;
                            }
                        }
                        else
                        {
                            RespondData = new RespondData() {errorCode = ErrorCode.FailedToDecryptServerPayload};
                        }
                    }
                    else
                    {
                        RespondData = JsonUtility.FromJson<RespondData>(response.DataAsText);
                    }
                }
                else
                {
                    RespondData = new RespondData() {errorCode = ErrorCode.Timeout};
                }

                isSending = false;
            });
            request.SetForm(postForm);
            request.Timeout = TimeSpan.FromSeconds(Timeout);
            request.Send();
        }
        
        public static byte[] Decompress(byte[] input)
        {
            using (var source = new MemoryStream(input))
            using (var zip = new GZipStream(source, CompressionMode.Decompress))
            using (var output = new MemoryStream())
            {
                zip.CopyTo(output);
                return output.ToArray();
            }
        }
    }
}

File: Johnny/encryption-final/src/tests/ApiClient.ts
import {IUser} from '../entities/IUser';
import {EncryptionService} from '../providers/encryption.service';
import {IRespondData} from '../entities/IRespondData';
import {ApiError} from '../errors/ApiError';
import {ErrorCode} from '../states/ErrorCode';
import {ICard} from '../entities/ICard';
import {IRequestData} from '../entities/IRequestData';
import axios from 'axios';
import * as crypto from 'crypto';
import * as zlib from 'zlib';
import {unzip} from "zlib";

interface ILastRequest {
    endPoint: string;
    data: IRequestData;
}

export class ApiClient {
    // services
    private _encryptionService: EncryptionService;
    public nonce: number = 0;
    public timeout: number;
    public version: string = '1.0';
    public versionKey: string = 'version-key';
    public lastRequest: ILastRequest | null;
    public localSendInvalidRequestData = false;
    public remoteSendInvalidPayload = false;
    public remoteTimeout = false;
    public localSendInvalidSignBase64 = false;

    // for testing
    public triggerInvalidServerPayload = false;
    public triggerInvalidSignature = false;
    public triggerInvalidClientPayload = false;
    public triggerTimeout = false;

    // return from server
    public user: IUser;
    public cards: ICard[];
    public receivedTimestamp: number;
    public serverTimestamp: number;
    public token: string;

    constructor(public readonly host: string, public readonly deviceId: string, requestTimeout: number) {
        this._encryptionService = new EncryptionService();
        this.timeout = requestTimeout;
        this.nonce = Math.random() * (1 << 30) | 0;
    }

    public get currentTimestamp(): number {
        return +new Date() / 1000 | 0;
    }

    public get predictedServerTimestamp(): number {
        const timestamp = this.currentTimestamp;
        return this.serverTimestamp + (timestamp - this.receivedTimestamp);
    }

    public async login() {
        const data = {deviceId: this.deviceId};
        const respondData = await this._request('/users/login', data);
        this.token = respondData.token;
    }

    public async listCards() {
        const data = {};
        const respondData = await this._request('/cards/list', data);
        this.cards = respondData.body.cards;
        return this.cards;
    }

    public async createCard(monsterId: number) {
        const data = {monsterId};
        const respondData = await this._request('/cards/create', data);
        this.cards = respondData.body.cards;
        return respondData;
    }

    public async updateCard(cardId: number, exp: number) {
        const data = {cardId, exp};
        const respondData = await this._request('/cards/update', data);
        this.cards = respondData.body.cards;
        return respondData;
    }

    public async deleteCard(cardId: number) {
        const data = {cardId};
        const respondData = await this._request('/cards/delete', data);
        this.cards = respondData.body.cards;
        return respondData;
    }

    public get canResend(): boolean {
        return !!(this.lastRequest);
    }

    public async resend(): Promise<IRespondData> {
        if (this.canResend) {
            return await this._request(this.lastRequest.endPoint, this.lastRequest.data);
        }
    }

    private async _request(endPoint: string, requestData: IRequestData): Promise<IRespondData> {
        requestData.version = this.version;
        requestData.cacheKey = Math.random().toString();
        requestData.timestamp = this.currentTimestamp;
        requestData.session = this.user ? this.user.session : '';
        this.receivedTimestamp = this.currentTimestamp;

        let nonceBuffer = this._encryptionService.generateNonce(this.nonce);
        this.nonce++;
        let payload;
        if(this.localSendInvalidRequestData){
            payload = this._encryptionService.encrypt(Buffer.from(requestData.toString()), nonceBuffer);
        }else{
            payload = this._encryptionService.encrypt(Buffer.from(JSON.stringify(requestData)), nonceBuffer);
        }
        let payloadBase64 = payload.toString('base64');
        let payloadMd5 = crypto.createHash('md5').update(payload).digest();
        let signMessage = Buffer.concat([nonceBuffer, payloadMd5]);
        const signedBase64 = this._encryptionService.signMessage(signMessage).toString('base64');

        if(this.localSendInvalidSignBase64){
            throw new ApiError(ErrorCode.InvalidSign, '', <IRespondData> null);
        }

        let response;
        if(this.remoteTimeout){
            axios.defaults.timeout = this.timeout;
        }
        response = await axios.request({
            method: 'post',
            url: this.host + endPoint,
            responseType: 'arraybuffer',
            data: {
                payloadBase64: payloadBase64,
                nonce: nonceBuffer,
                j:1
            },
            params: {
                signedBase64: signedBase64,
                token: this.token,
                remoteSendInvalidPayload: this.remoteSendInvalidPayload,
                remoteTimeout: this.remoteTimeout

            },
            timeout: this.timeout,
            // transformResponse: (r: any) => {
            //     return r;
            // }
        }).catch (e => {
            requestData.cacheKey = Math.random().toString();
            this.lastRequest = {endPoint: endPoint, data: requestData};
            throw new ApiError(ErrorCode.Timeout, '', <IRespondData> null);
        });

        if(response.data.toString().includes('errorCode')){
            let errorResponse = <IRespondData> JSON.parse(response.data.toString());
            if(errorResponse.errorCode == ErrorCode.FailedToDecryptClientPayload){
               this.localSendInvalidRequestData = true;
            }else if(errorResponse.errorCode == ErrorCode.Timeout){
               this.remoteTimeout = true;
            }
            throw new ApiError(errorResponse.errorCode, '', errorResponse);
        }else{
            let decryptedData = this._encryptionService.decrypt(response.data, nonceBuffer);
            response = await new Promise((resolve, reject) => {
                zlib.gunzip(decryptedData, (err, res) => {
                    if(res == undefined){
                        reject(ErrorCode.FailedToDecryptServerPayload);
                        return new ApiError(ErrorCode.FailedToDecryptServerPayload, '', <IRespondData> null);
                    }else{
                        if(this.lastRequest){
                            this.lastRequest = <ILastRequest> null;
                        }
                        let unzipedData = <IRespondData> JSON.parse(res.toString());
                        unzipedData.user = <IUser> unzipedData.user;
                        this.user = unzipedData.user;
                        if(Math.abs(this.serverTimestamp - this.receivedTimestamp) > 3600){
                            reject(ErrorCode.InvalidTimestamp);
                            return new ApiError(ErrorCode.InvalidTimestamp, '', <IRespondData> null);
                        }
                        this.serverTimestamp = unzipedData.timestamp;
                        if(requestData.cacheKey){
                            unzipedData.isCache = true;
                        }
                        resolve(unzipedData);
                        // console.log(unzipedData);
                    }
                });

            }).catch(e => {
                throw new ApiError(e, '', <IRespondData> null);
            }).then(value => { return value; });
        }

        return response;
    }
}
